### PR TITLE
CategoricalAccessor.categorical removed in 0.24.0rc1 #24751

### LINKED ISSUE #24751: CategoricalAccessor.categorical removed in 0.24.0rc1
Previously, CategoricalAccessor set `.categorical` in `__init__`. I don't think this was intended to be the way to access things like `.ordered`, but it wasn't prefixed with an underscore, so we should deprecate before removing

0.23.4

```python
In [11]: pd.Series(['a', 'b'], dtype='category').cat.categorical.ordered
Out[11]: False
```

0.24.0rc1

```python
In [3]: pd.Series(['a', 'b'], dtype='category').cat.categorical.ordered
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-e5c7f0300480> in <module>
----> 1 pd.Series(['a', 'b'], dtype='category').cat.categorical.ordered

AttributeError: 'CategoricalAccessor' object has no attribute 'categorical'
```

We can deprecate with

```python
def categorical(self):
    warnings.warn(...)
    return self._parent
```

in https://github.com/pandas-dev/pandas/blob/33f91d8f9f2e84f2b5f3ac3f0481b691c977c427/pandas/core/arrays/categorical.py#L2472

We should do the same for `.index` and `.name`.

### PR DESCRIPTION
**Pull request by benjaminr**
- [ ] closes #24751 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
see my comments on the issue.

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -1309,6 +1309,7 @@ Deprecations
 - :meth:`Series.clip_lower`, :meth:`Series.clip_upper`, :meth:`DataFrame.clip_lower` and :meth:`DataFrame.clip_upper` are deprecated and will be removed in a future version. Use ``Series.clip(lower=threshold)``, ``Series.clip(upper=threshold)`` and the equivalent ``DataFrame`` methods (:issue:`24203`)
 - :meth:`Series.nonzero` is deprecated and will be removed in a future version (:issue:`18262`)
 - Passing an integer to :meth:`Series.fillna` and :meth:`DataFrame.fillna` with ``timedelta64[ns]`` dtypes is deprecated, will raise ``TypeError`` in a future version.  Use ``obj.fillna(pd.Timedelta(...))`` instead (:issue:`24694`)
+- The ``name``, ``index`` and ``categorical`` attributes of ``CategoricalAccessor`` have been removed in 0.24.0rc1, but should be deprecated for the time being, raising an appropriate warning (:issue:`24751`).
```
[the remark this answers is not in the document]
Comment (benjaminr) [reply doc/source/whatsnew/v0.24.0.rst]: Okay, shall do. And noted your other comment about it this change potentially being entirely unnecessary.

**Comment by benjaminr** [comment]
Given @jreback's suggestion that this is unnecessary, I will close this.

File: doc/source/whatsnew/v0.24.0.rst [hunk 2]
```diff
@@ -1309,6 +1309,7 @@ Deprecations
 - :meth:`Series.clip_lower`, :meth:`Series.clip_upper`, :meth:`DataFrame.clip_lower` and :meth:`DataFrame.clip_upper` are deprecated and will be removed in a future version. Use ``Series.clip(lower=threshold)``, ``Series.clip(upper=threshold)`` and the equivalent ``DataFrame`` methods (:issue:`24203`)
 - :meth:`Series.nonzero` is deprecated and will be removed in a future version (:issue:`18262`)
 - Passing an integer to :meth:`Series.fillna` and :meth:`DataFrame.fillna` with ``timedelta64[ns]`` dtypes is deprecated, will raise ``TypeError`` in a future version.  Use ``obj.fillna(pd.Timedelta(...))`` instead (:issue:`24694`)
+- :attr:``CategoricalAccessor.categorical``, :attr:``CategoricalAccessor.name`` and :attr:``CategoricalAccessor.index`` have been deprecated and replaced by private :attr:``CategoricalAccessor._parent``, :attr:``CategoricalAccessor._name`` and :attr:``CategoricalAccessor._index``, respectively. (:issue:`24751`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: I don't think any of these links will work, as they aren't part of the API docs. We also don't want to direct people to use private methods. Rather, I'd say

    ``Series.cat.categorical``, ``Series.cat.name`` and ``Sersies.cat.index`` have been deprecated. Use
    the attributes on ``Series.cat`` or ``Series`` directly.

File: pandas/core/arrays/categorical.py [hunk 3]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
         if res is not None:
             return Series(res, index=self.index, name=self.name)
 
+    @property
+    def categorical(self):
+        warn("s.cat.categorical has been deprecated",
+             DeprecationWarning,
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: FutureWarning.

File: pandas/core/arrays/categorical.py [hunk 4]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
         if res is not None:
             return Series(res, index=self.index, name=self.name)
 
+    @property
+    def categorical(self):
+        warn("s.cat.categorical has been deprecated",
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Say `'Series.cat.categorical'`, rather than `'s'`. Also, say what to do instead. Something like `Use the attributes on 'Series.cat' directly instead.`.

File: pandas/core/arrays/categorical.py [hunk 5]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
...
+             DeprecationWarning,
+             stacklevel=2)
+        return self._parent
+
+    @property
+    def name(self):
+        warn("s.cat.name has been deprecated",
+             DeprecationWarning,
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: FutureWarning.

File: pandas/core/arrays/categorical.py [hunk 6]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
...
+    @property
+    def categorical(self):
+        warn("s.cat.categorical has been deprecated",
+             DeprecationWarning,
+             stacklevel=2)
+        return self._parent
+
+    @property
+    def name(self):
+        warn("s.cat.name has been deprecated",
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: `s` -> `Series`. Say use `Series.name` instead.

File: pandas/core/arrays/categorical.py [hunk 7]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
...
+             DeprecationWarning,
+             stacklevel=2)
+        return self._name
+
+    @property
+    def index(self):
+        warn("s.cat.index has been deprecated",
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: `s` -> `Series`. Say "Use Series.index instead"

File: pandas/core/arrays/categorical.py [hunk 8]
```diff
@@ -2529,6 +2529,26 @@ def _delegate_method(self, name, *args, **kwargs):
...
+             DeprecationWarning,
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: FutureWarning

File: pandas/tests/arrays/categorical/test_warnings.py [hunk 9]
```diff
@@ -16,3 +17,15 @@ def test_tab_complete_warning(self, ip):
         with tm.assert_produces_warning(None):
             with provisionalcompleter('ignore'):
                 list(ip.Completer.completions('c.', 1))
+
+    def test_CategoricalAccessor_categorical_deprecation(object):
+        with tm.assert_produces_warning(DeprecationWarning):
```
Comment (TomAugspurger) [inline pandas/tests/arrays/categorical/test_warnings.py]: This will need to be updated to FutureWarnings

**Review by TomAugspurger — COMMENTED** [review]
May need to find places where we use `.cat.name` or `.cat.index` internally (if any).

**Comment by TomAugspurger** [comment]
You'll also need to update uses of `self.index` to `self._index` or `self._parent.index`. I didn't realize these were public on master. It's probably best make these private as well (as you've done), just have to make sure we don't use the deprecated version internally.

File: pandas/tests/series/test_api.py [hunk 10]
```diff
@@ -245,10 +245,11 @@ def test_tab_completion(self):
 
     def test_tab_completion_with_categorical(self):
         # test the tab completion display
-        ok_for_cat = ['categories', 'codes', 'ordered', 'set_categories',
```
Comment (TomAugspurger) [inline pandas/tests/series/test_api.py]: Why was this changed needed? IIUC, then name, index, and categorical should *not* be included here anymore.

Comment (benjaminr) [reply pandas/tests/series/test_api.py]: I wasn't sure either, when I was running the tests locally, this was raising an AssertionError. Can remove it and see how the tests do.

Comment (TomAugspurger) [reply pandas/tests/series/test_api.py]: Sorry, you will need the change you had originally, not sure what I was thinking of last night.

These are going to be in `dir(Series.cat)`, because you're adding properties for the deprecation. I think they weren't there original, because those attributes were set in the `__init__`, and IPython was analyzing the class before that was called.

Could you add a note in the property `CategoricalAccessor.categorical` saying that this test will need to be updated too when the deprecation is enforced?

Comment (benjaminr) [reply pandas/tests/series/test_api.py]: Ah no worries.

When you say note on that property, is a comment sufficient, or do you mean in the deprecation FutureWarning message?

**Comment by TomAugspurger** [comment]
Just a comment.

File: pandas/core/arrays/categorical.py [hunk 11]
```diff
@@ -2520,15 +2520,53 @@ def codes(self):
...
         from pandas import Series
         method = getattr(self._parent, name)
         res = method(*args, **kwargs)
         if res is not None:
-            return Series(res, index=self.index, name=self.name)
+            return Series(res, index=self._index, name=self._name)
 
+    @property
+    def categorical(self):
+        """
```
Comment (TomAugspurger) [inline pandas/core/arrays/categorical.py]: Can you change these to regular comments. Users see these docstrings, and this comment is for developers.

Comment (benjaminr) [reply pandas/core/arrays/categorical.py]: Yeah apologies - bit of an oversight.

**Review by TomAugspurger — APPROVED** [review]
Ping on green.

**Review by jreback — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @benjaminr!
